File: src/main/java/ru/octol1ttle/flightassistant/FAKeyBindings.java
package ru.octol1ttle.flightassistant;

import java.util.ArrayList;
import java.util.List;
import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingHelper;
import net.minecraft.client.option.KeyBinding;
import net.minecraft.client.util.InputUtil;
import org.lwjgl.glfw.GLFW;
import ru.octol1ttle.flightassistant.computers.impl.autoflight.AutoFlightController;
import ru.octol1ttle.flightassistant.computers.impl.autoflight.ThrustController;
import ru.octol1ttle.flightassistant.computers.impl.safety.AlertController;
import ru.octol1ttle.flightassistant.registries.ComputerRegistry;

public class FAKeyBindings {
    private final List<KeyBinding> keyBindings = new ArrayList<>();
    
    private final KeyBinding toggleFlightDirectors;
    private final KeyBinding toggleAutoThrust;
    private final KeyBinding toggleAutoPilot;

    private final KeyBinding hideAlert;
    private final KeyBinding recallAlert;

    private final KeyBinding setIdle;
    private final KeyBinding decreaseThrust;
    private final KeyBinding increaseThrust;
    private final KeyBinding setToga;

    public FAKeyBindings() {
        toggleFlightDirectors = addKeyBinding("key.flightassistant.toggle_flight_directors", GLFW.GLFW_KEY_KP_1);
        toggleAutoThrust = addKeyBinding("key.flightassistant.toggle_auto_thrust", GLFW.GLFW_KEY_KP_2);
        toggleAutoPilot = addKeyBinding("key.flightassistant.toggle_auto_pilot", GLFW.GLFW_KEY_KP_3);

        hideAlert = addKeyBinding("key.flightassistant.hide_alert", GLFW.GLFW_KEY_KP_0);
        recallAlert = addKeyBinding("key.flightassistant.recall_alert", GLFW.GLFW_KEY_KP_DECIMAL);

        setIdle = addKeyBinding("key.flightassistant.set_idle", GLFW.GLFW_KEY_LEFT);
        decreaseThrust = addKeyBinding("key.flightassistant.decrease_thrust", GLFW.GLFW_KEY_DOWN);
        increaseThrust = addKeyBinding("key.flightassistant.increase_thrust", GLFW.GLFW_KEY_UP);
        setToga = addKeyBinding("key.flightassistant.set_toga", GLFW.GLFW_KEY_RIGHT);
    }
    
    private KeyBinding addKeyBinding(String translationKey, int code) {
        KeyBinding keyBinding = new KeyBinding(translationKey, InputUtil.Type.KEYSYM, code, "mod.flightassistant");
        keyBindings.add(keyBinding);
        return keyBinding;
    }

    public void registerAll() {
        for (KeyBinding keyBinding : keyBindings) {
            KeyBindingHelper.registerKeyBinding(keyBinding);
        }
    }

    public void tick() {
        if (!ComputerRegistry.isFaulted(AlertController.class)) {
            AlertController alert = ComputerRegistry.resolve(AlertController.class);
            while (hideAlert.wasPressed()) {
                alert.hide();
            }
            while (recallAlert.wasPressed()) {
                alert.recall();
            }
        }

        boolean disconnectAutoThrust = false;
        if (!ComputerRegistry.isFaulted(ThrustController.class)) {
            ThrustController thrust = ComputerRegistry.resolve(ThrustController.class);

            while (setIdle.wasPressed()) {
                thrust.setThrust(0.0f);
                disconnectAutoThrust = true;
            }
            while (setToga.wasPressed()) {
                thrust.setThrust(1.0f);
                disconnectAutoThrust = true;
            }

            while (decreaseThrust.wasPressed()) {
                thrust.addThrustTick(-1.0f);
                disconnectAutoThrust = true;
            }
            while (increaseThrust.wasPressed()) {
                thrust.addThrustTick(1.0f);
                disconnectAutoThrust = true;
            }
        }

        if (!ComputerRegistry.isFaulted(AutoFlightController.class)) {
            AutoFlightController autoflight = ComputerRegistry.resolve(AutoFlightController.class);

            if (disconnectAutoThrust) {
                autoflight.disconnectAutoThrust(true);
            }

            while (toggleFlightDirectors.wasPressed()) {
                autoflight.flightDirectorsEnabled = !autoflight.flightDirectorsEnabled;
            }

            while (toggleAutoThrust.wasPressed()) {
                if (!autoflight.autoThrustEnabled) {
                    autoflight.autoThrustEnabled = true;
                } else {
                    autoflight.disconnectAutoThrust(false);
                }
            }

            while (toggleAutoPilot.wasPressed()) {
                if (!autoflight.autoPilotEnabled) {
                    autoflight.autoPilotEnabled = true;
                } else {
                    autoflight.disconnectAutopilot(false);
                }
            }
        }
    }
}

File: src/main/kotlin/ru/octol1ttle/flightassistant/FAKeyBindings.kt
package ru.octol1ttle.flightassistant

import net.minecraft.client.option.KeyBinding
import net.minecraft.client.util.InputUtil
import org.lwjgl.glfw.GLFW
import ru.octol1ttle.flightassistant.FlightAssistant.mc
import ru.octol1ttle.flightassistant.api.computer.ComputerView
import ru.octol1ttle.flightassistant.api.util.FATickCounter
import ru.octol1ttle.flightassistant.screen.FlightSetupScreen

object FAKeyBindings {
    internal val keyBindings: MutableList<KeyBinding> = ArrayList()

    private lateinit var openFlightSetup: KeyBinding

    private lateinit var autopilotDisconnect: KeyBinding
    private lateinit var manualPitchOverride: KeyBinding

    private lateinit var hideCurrentAlert: KeyBinding
    private lateinit var showHiddenAlert: KeyBinding

    private lateinit var setIdle: KeyBinding
    private lateinit var decreaseThrust: KeyBinding
    private lateinit var increaseThrust: KeyBinding
    private lateinit var setToga: KeyBinding

    fun setup() {
        openFlightSetup = addKeyBinding("keys.flightassistant.open_flight_setup", GLFW.GLFW_KEY_KP_ENTER)

        autopilotDisconnect = addKeyBinding("keys.flightassistant.autopilot_disconnect", GLFW.GLFW_KEY_CAPS_LOCK)
        manualPitchOverride = addKeyBinding("keys.flightassistant.manual_pitch_override", GLFW.GLFW_KEY_LEFT_ALT)

        hideCurrentAlert = addKeyBinding("keys.flightassistant.hide_current_alert", GLFW.GLFW_KEY_KP_0)
        showHiddenAlert = addKeyBinding("keys.flightassistant.show_hidden_alert", GLFW.GLFW_KEY_KP_DECIMAL)

        setIdle = addKeyBinding("keys.flightassistant.set_idle", GLFW.GLFW_KEY_LEFT)
        decreaseThrust = addKeyBinding("keys.flightassistant.decrease_thrust", GLFW.GLFW_KEY_DOWN)
        increaseThrust = addKeyBinding("keys.flightassistant.increase_thrust", GLFW.GLFW_KEY_UP)
        setToga = addKeyBinding("keys.flightassistant.set_toga", GLFW.GLFW_KEY_RIGHT)
    }

    private fun addKeyBinding(translationKey: String, code: Int, type: InputUtil.Type = InputUtil.Type.KEYSYM): KeyBinding {
        val keyBinding = KeyBinding(translationKey, type, code, "keys.flightassistant")
        keyBindings.add(keyBinding)
        return keyBinding
    }

    fun checkPressed(computers: ComputerView) {
        while (openFlightSetup.wasPressed()) {
            mc.execute {
                mc.setScreen(FlightSetupScreen())
            }
        }

        while (autopilotDisconnect.wasPressed()) {
            computers.automations.setAutoPilot(false, alert = false)
        }
        computers.pitch.manualOverride = manualPitchOverride.isPressed

        while (hideCurrentAlert.wasPressed()) {
            computers.alert.hideCurrentAlert()
        }
        while (showHiddenAlert.wasPressed()) {
            computers.alert.showHiddenAlert()
        }

        while (setIdle.wasPressed()) {
            computers.thrust.setTarget(0.0f)
        }
        while (setToga.wasPressed()) {
            computers.thrust.setTarget(1.0f)
        }
        while (decreaseThrust.wasPressed()) {
            computers.thrust.setTarget((computers.thrust.current - FATickCounter.timePassed / 3).coerceIn(-1.0f..1.0f))
        }
        while (increaseThrust.wasPressed()) {
            computers.thrust.setTarget((computers.thrust.current + FATickCounter.timePassed / 3).coerceIn(-1.0f..1.0f))
        }
    }
}
